py/server/tests/test_sql.py

```diff
@@ -0,0 +1,127 @@
+
+    def test_dry_run_global_explicit(self):
+        result_spec = sql.eval(
+            "SELECT * FROM my_global_table",
+            dry_run=True,
+            globals={"my_global_table": some_global_table},
+        )
+        self.assertTrue(isinstance(result_spec, jpy.JType))
+        # Might be nice to extend jpy like this in the future?
+        # self.assertTrue(isinstance(result_spec, _JTableSpec))
+        self.assertTrue(_JTableSpec.jclass.isInstance(result_spec))
+
+    def test_dry_run_local_implicit(self):
+        # noinspection PyUnusedLocal
+        test_table = read_csv("tests/data/test_table.csv")
+        result_spec = sql.eval("SELECT * FROM test_table LIMIT 7", dry_run=True)
+        self.assertTrue(isinstance(result_spec, jpy.JType))
+        # Might be nice to extend jpy like this in the future?
+        # self.assertTrue(isinstance(result_spec, _JTableSpec))
+        self.assertTrue(_JTableSpec.jclass.isInstance(result_spec))
+
+    def test_dry_run_local_explicit(self):
+        # noinspection PyUnusedLocal
+        test_table = read_csv("tests/data/test_table.csv")
+        result_spec = sql.eval(
+            "SELECT * FROM my_table LIMIT 8",
+            dry_run=True,
+            locals={"my_table": test_table},
+        )
+        self.assertTrue(isinstance(result_spec, jpy.JType))
+        # Might be nice to extend jpy like this in the future?
+        # self.assertTrue(isinstance(result_spec, _JTableSpec))
+        self.assertTrue(_JTableSpec.jclass.isInstance(result_spec))
+
+    def test_local_takes_precedence(self):
+        # noinspection PyUnusedLocal,PyShadowingNames
+        some_global_table = empty_table(13)
+        result_table = sql.eval("SELECT * FROM some_global_table")
+        self.assertEqual(result_table.size, 13)
+
+    def test_no_implicit_global_if_local_explicit(self):
+        test_table = read_csv("tests/data/test_table.csv")
+        with self.assertRaises(DHError):
+            sql.eval(
+                "SELECT * FROM my_global_table",
+                locals={"test_table": test_table},
+            )
+
+    def test_no_implicit_local_if_global_explicit(self):
+        # noinspection PyUnusedLocal
+        test_table = read_csv("tests/data/test_table.csv")
+        with self.assertRaises(DHError):
+            sql.eval(
+                "SELECT * FROM test_table",
+                globals={"my_global_table": some_global_table},
+            )
+
+    def test_nested_non_local(self):
+        # noinspection PyUnusedLocal
+        test_table = read_csv("tests/data/test_table.csv")
+
+        def do_sql():
+            nonlocal test_table
+            return sql.eval("SELECT * FROM test_table LIMIT 3")
+
+        result_table = do_sql()
+        self.assertEqual(result_table.size, 3)
+
+    def test_current_timestamp(self):
+        result_table = sql.eval("SELECT CURRENT_TIMESTAMP")
+        self.assertEqual(result_table.size, 1)
+
+
```

Comment:
We tested a few other scoping scenarios for the table methods.  Consider testing some of those here as well.  See for example the stuff around this line in the test file.
https://github.com/deephaven/deephaven-core/blob/main/py/server/tests/test_table.py#L730

Reply:
The scoping for SQL is a bit more limited, as it's limited to only referencing table names. Custom functions are on the roadmap, and we'd be able to get more tests of this kind then.

Reply:
The test cases I'm referring to as examples make sure that query scope is properly obtained from the correct Python scope.

Reply:
I've added a new test, but it's not really testing anything new - is there another specific SQL test you want to see?